extensions/amp-accordion/1.0/accordion.js

```diff
@@ -0,0 +1,176 @@
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {omit} from '../../../src/utils/object';
+import {sequentialIdGenerator} from '../../../src/utils/id-generator';
+import {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from '../../../src/preact';
+
+const AccordionContext = Preact.createContext(
+  /** @type {AccordionDef.ContextProps} */ ({})
+);
+
+/** @type {!Object<string, boolean>} */
+const EMPTY_EXPANDED_SET = {};
+
+const generateSectionId = sequentialIdGenerator();
+
+/**
+ * @param {!AccordionDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function Accordion({
+  as: Comp = 'section',
+  expandSingleSection = false,
+  children,
+  ...rest
+}) {
+  const [expandedSet, setExpandedSet] = useState(EMPTY_EXPANDED_SET);
+
+  useEffect(() => {
+    if (!expandSingleSection) {
+      return;
+    }
+    setExpandedSet((expandedSet) => {
+      const newExpandedSet = {};
+      let expanded = 0;
+      for (const k in expandedSet) {
+        newExpandedSet[k] = expandedSet[k] && expanded++ == 0;
+      }
+      return newExpandedSet;
+    });
+  }, [expandSingleSection]);
+
+  const registerSection = useCallback((id) => {
+    return () => setExpandedSet((expandedSet) => omit(expandedSet, id));
+  }, []);
+
+  const toggleExpanded = useCallback(
+    (id, defaultExpanded) => {
+      setExpandedSet((expandedSet) => {
+        const newValue = !isExpanded(
+          id,
+          defaultExpanded,
+          expandedSet,
+          expandSingleSection
+        );
+        let newExpandedSet;
+        if (newValue && expandSingleSection) {
+          newExpandedSet = {[id]: newValue};
+          for (const k in expandedSet) {
+            if (k != id) {
+              newExpandedSet[k] = false;
+            }
+          }
+        } else {
+          newExpandedSet = {...expandedSet, [id]: newValue};
+        }
+        return newExpandedSet;
+      });
+    },
+    [expandSingleSection]
+  );
+
+  const context = useMemo(
+    () =>
+      /** @type {!AccordionDef.ContextProps} */ ({
+        registerSection,
+        toggleExpanded,
+        isExpanded: (id, defaultExpanded) =>
+          isExpanded(id, defaultExpanded, expandedSet, expandSingleSection),
+      }),
+    [expandedSet, expandSingleSection, registerSection, toggleExpanded]
+  );
+
+  return (
+    <Comp {...rest}>
+      <AccordionContext.Provider value={context}>
+        {children}
+      </AccordionContext.Provider>
+    </Comp>
+  );
+}
+
+/**
+ * @param {string} id
+ * @param {boolean} defaultExpanded
+ * @param {!Object<string, boolean>} expandedSet
+ * @param {boolean} expandSingleSection
+ * @return {boolean}
+ */
+function isExpanded(id, defaultExpanded, expandedSet, expandSingleSection) {
+  const current = expandedSet[id];
+  if (current != null) {
+    return current;
+  }
+  const adjDefaultExpanded = expandSingleSection
```

Comment:
`adj` probably a bad name. I meant it to mean "adjusted", but that's still bad. Basically, here's what's going on here. If a new section is added to the accordion, whether it should be immediately expanded is based on `expandSingleSelection` mode:
* `expandSingleSelection == false` - we just do whatever the section itself wants. If it was defined with `expanded` prop/attr - we can use that.
* `expandSingleSelection == true` - we want to just use the section's defined `expanded`, but we can only do so if there are no other currently expanded sections. Otherwise we break the contract of "only one section can be expanded over time".

Maybe a better alternative would be to just use the `expanded` defined by the section, but collapse a previous expanded section?

Reply:
If a new section is added, it seems unintuitive that we would ignore its default expanded state if `expandSingleSelection == true` -- I'm leaning towards collapsing previously expanded section

Reply:
Makes sense. Reimplemented.